activerecord/test/cases/relation/or_test.rb

```diff
@@ -138,5 +138,15 @@ def test_or_with_scope_on_association
         author.top_posts.or(author.other_top_posts)
       end
     end
+
+    def test_or_with_multiple_joins
+      assert_nothing_raised do
+        Post.joins(:author).joins(:author).or(Post.joins(:author))
```

Comment:
If the purpose of this PR is to allow such like values for `or`, `joins` and `left_joins` are not only the case (e.g. `eager_load` etc).

Reply:
Thank you for your review.
I checked the methods in [STRUCTURAL_OR_METHODS](https://github.com/rails/rails/blob/9171763d197bdd814fafaff0278712b2aed4d737/activerecord/lib/active_record/relation/query_methods.rb#L1360-L1367).

```ruby
ActiveRecord::QueryMethods::STRUCTURAL_OR_METHODS
#=> [:includes, :eager_load, :preload, :select, :group, :order, :joins, :left_outer_joins, :optimizer_hints, :annotate, :limit, :offset, :lock, :readonly, :reordering, :reverse_order, :distinct, :create_with, :skip_query_cache, :from]
```

### includes, eager_load, preload, group, joins, left_outer_joins

I fixed.

### limit, offset, lock, readonly, reordering, reverse_order, distinct, create_with, skip_query_cache

These use `=` instead of `+=` and do not need to be fixed.

### optimizer_hints

It uses `|=` and do not need to be fixed.

### annotate

An error occurs if the SQL comment is different, but I think it should not be an error.
I want to receive the opinion from Rails committer before the changes.

```ruby
assert_equal 'SELECT "posts".* FROM "posts" /* foo */ /* bar */',
  Post.annotate('foo').or(Post.annotate('bar')).to_sql
#=> ArgumentError: Relation passed to #or must be structurally compatible. Incompatible values: [:annotate]
```

<details>
<summary>test_script.rb</summary>

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "6.0.0"
  # gem "activerecord", path: '/Users/sinsoku/ghq/github.com/rails/rails/activerecord'
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table(:posts, force: true)
end

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def test_or_with_annotate
    assert_equal 'SELECT "posts".* FROM "posts" /* foo */',
      Post.annotate('foo').or(Post.annotate('foo')).to_sql
    # Success

    assert_equal 'SELECT "posts".* FROM "posts" /* foo */ /* bar */',
      Post.annotate('foo').or(Post.annotate('bar')).to_sql
    #=> ArgumentError: Relation passed to #or must be structurally compatible. Incompatible values: [:annotate]
  end
end
```
</details>

### select, order

`select` and `order` have the same problem.

```ruby
Post.select(:id).select(:id).values[:select].size
#=> 2
Post.order(:id).order(:id).values[:order].size
#=> 2
```

However, the use case used with the `or` method does not come to mind.
Should I fix these?

### from

The from method uses `=`, but the same value causes an error.

```ruby
Post.from('users').or(User.from('users'))
#=> ArgumentError (Relation passed to #or must be structurally compatible. Incompatible values: [:from])
```

I can't imagine a use case that would be used with or.
Should I fix `from` to work with `or`?

Reply:
* `annotate`

If we will change to merge condition not only `where` and `having` but also `annotate`, it is notable change and also the doc should be updated. I prefer to discuss about that in another PR.

https://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-or

* `select`, `order`

`select` and `order` is order dependent (e.g. dedup `select_values = [:id, "other_table.id", :id]` to `select_values = [:id, "other_table.id"]` will change public behavior), so it should be kept as it is from that point of view.

* `from`

This is probably easy to fix with like this:

```diff
diff --git a/activerecord/lib/active_record/relation/from_clause.rb b/activerecord/lib/active_record/relation/from_clause.rb
index 93b129ce99..222b1c70f0 100644
--- a/activerecord/lib/active_record/relation/from_clause.rb
+++ b/activerecord/lib/active_record/relation/from_clause.rb
@@ -21,6 +21,10 @@ def empty?
       def self.empty
         @empty ||= new(nil, nil).freeze
       end
+
+      def ==(other)
+        self.class == other.class && value == other.value && name == other.name
+      end
     end
   end
 end
```

Reply:
Thanks.
I fixed the `or` method to work with the `from` clause with the same value.

I will create an another PR  that changes the behavior when used with `annotate`.